src/node_api.cc

```diff
@@ -627,28 +715,22 @@ NAPI_NO_RETURN void napi_fatal_error(const char* location,
 }
 
 napi_status napi_open_callback_scope(napi_env env,
-                                     napi_value resource_object,
+                                     napi_value /** ignored */,
                                      napi_async_context async_context_handle,
                                      napi_callback_scope* result) {
   // Omit NAPI_PREAMBLE and GET_RETURN_STATUS because V8 calls here cannot throw
   // JS exceptions.
   CHECK_ENV(env);
   CHECK_ARG(env, result);
 
-  v8::Local<v8::Context> context = env->context();
-
-  node::async_context* node_async_context =
-      reinterpret_cast<node::async_context*>(async_context_handle);
-
-  v8::Local<v8::Object> resource;
-  CHECK_TO_OBJECT(env, context, resource, resource_object);
+  v8impl::AsyncContext* node_async_context =
+      reinterpret_cast<v8impl::AsyncContext*>(async_context_handle);
+  if (node_async_context->lost_reference()) {
+    return napi_set_last_error(env, napi_async_out_of_scope);
+  }
```

Comment:
I still wonder if we can make this less breaking for existing code. Particularly since Async Hooks may not be enabled. If the resource has been "lost" we can document as a programmer error (not keeping resource alive while it is needed)  which will result in loss of context, but at the same time not return an error ( can resource be passed as NULL in this case?).  This would allow existing code to continue to run, with only async hooks/context propagation being broken (although maybe no worse than it was before?)

Reply:
PR Updated to remove the assertions and trying to recover from the reference lost and keep `async_hooks.executionAsyncResource` working (might not be the correct result, but it's not worse than status quo). PTAL :)